dev/spark_overview.md

```diff
@@ -0,0 +1,452 @@
+
+spark = SparkSession.getActiveSession()
+spark.conf.set("spark.sql.shuffle.partitions", 500)
+```
+
+## Creating a SparkSession with dsgrid
+Ensure that the dsgrid software uses the cluster with optimized settings. Most importantly,
+don't create a session in local mode, bypassing the cluster. You can verify that you are
+using the correct settings in the Spark UI.
+
+### Through Spark scripts
+This is the recommended procedure because you can customize any setting.
+
+Suppose that you want to run a query with a CLI command and use custom settings.
+This example assumes that you are on the HPC compute node that is the Spark master.
+
+**Note**: `spark-submit` needs to be able to detect that the script is Python, and so you must
+substitute `dsgrid-cli.py` for the usual `dsgrid`. You also need to specify its full path.
+
+```
+$ spark-submit --master spark://$(hostname):7077 \
+    --conf spark.sql.shuffle.partitions=500 \
+    $(which dsgrid-cli.py) \
+    query project run \
+    --offline \
+    --registry-path=/scratch/${USER}/.dsgrid-registry \
+    query.json
+```
+
+Here is a similar example if you want to run code interactively in ipython or jupyter.
+
+```
+$ pyspark --master spark://$(hostname):7077 --conf spark.sql.shuffle.partitions=500
+```
+
+In both cases the Spark scripts will create a session and make it available to the Python
+process. dsgrid will connect to it.
+
+### At dsgrid runtime
+dsgrid will attempt to connect to an existing cluster through the `SPARK_CLUSTER` environment
+variable. If you want to choose this route:
+```
+$ export SPARK_CLUSTER=spark://$(hostname):7077
+$ dsgrid query project run --offline --registry-path=/scratch/${USER}/.dsgrid-registry query.json
+```
+
+The upside of this approach is that it is a bit simpler. The downside is that you cannot configure
+settings like `spark.driver.memory`, which, as stated earlier, must be set before the JVM is
+created.
+
+
+## Installing a Spark Standalone Cluster on your laptop
+Download your desired version from https://spark.apache.org/downloads.html and extract it on
+your system.
+
+**Note**: Choose the version that is set in `dsgrid/setup.py` for `pyspark`. At least the minor
+versions must match.
+
+Here is an example:
+```
+$ cd
+$ wget https://dlcdn.apache.org/spark/spark-3.3.1/spark-3.3.1-bin-hadoop3.tgz
+$ tar -xzf spark-3.3.1-bin-hadoop3.tgz && rm spark-3.3.1-bin-hadoop3.tgz
+```
+
+The full instructions to create a cluster are at
+http://spark.apache.org/docs/latest/spark-standalone.html. The rest of this section documents a
+limited set that should work on your system.
+
+### Set environment variables
+
+```
+# Consider adding these to your shell rc file (.bashrc or .zshrc, for example).
+# so that they are always set.
+$ export SPARK_HOME=~/spark-3.3.1-bin-hadoop3
+$ export PATH=$PATH:$SPARK_HOME/sbin
+```
+
+**Warning**: Setting `SPARK_HOME` will affect `pyspark` operation in local mode.
+
+### Customize Spark configuration settings
```

Comment:
Is it worth sharing what the default value would be if blank? e.g., `spark.sql.shuffle.partitions` = 200

Reply:
That is mentioned elsewhere

Reply:
I know, but I think the defaults are more relevant and useful to have in this section here, and you only really talked about default for partitions later.